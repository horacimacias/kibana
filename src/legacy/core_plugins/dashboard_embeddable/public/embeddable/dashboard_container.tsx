/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import ReactDOM from 'react-dom';
import uuid from 'uuid';

import { I18nProvider } from '@kbn/i18n/react';
import { IndexPattern } from 'ui/index_patterns';

import { TimeRange } from 'ui/timefilter/time_history';
import {
  Container,
  ContainerInput,
  Embeddable,
  EmbeddableFactoryRegistry,
  EmbeddableInput,
  Filter,
  Query,
  RefreshConfig,
  ViewMode,
  EmbeddableOutput,
  EmbeddableInputMissingFromContainer,
  isErrorEmbeddable,
  EmbeddableFactory,
} from '../../../embeddable_api/public/index';

import { DASHBOARD_CONTAINER_TYPE } from './dashboard_container_factory';
import { createPanelState } from './panel';
import { DashboardPanelState } from './types';
import { DashboardViewport } from './viewport/dashboard_viewport';

export interface DashboardContainerInput extends ContainerInput {
  viewMode: ViewMode;
  filters: Filter[];
  query: Query;
  timeRange: TimeRange;
  refreshConfig?: RefreshConfig;
  expandedPanelId?: string;
  useMargins: boolean;
  title: string;
  description?: string;
  isFullScreenMode: boolean;
  panels: { [panelId: string]: DashboardPanelState };

  // Used to force a refresh of embeddables even if there were no other input state
  // changes.
  lastReloadRequestTime?: number;
}

export interface DashboardEmbeddableInput {
  filters: Filter[];
  query: Query;
  timeRange: TimeRange;
  refreshConfig?: RefreshConfig;
  viewMode: ViewMode;
  hidePanelTitles?: boolean;
  id: string;
  firstName: string;
}

export interface DashboardEmbeddableOutput extends EmbeddableOutput {
  indexPatterns?: IndexPattern[];
}

export type DashboardEmbeddable = Embeddable<
  DashboardEmbeddableInput & EmbeddableInput,
  DashboardEmbeddableOutput
>;

export class DashboardContainer extends Container<
  DashboardEmbeddableInput,
  DashboardEmbeddableOutput,
  DashboardContainerInput
> {
  constructor(
    initialInput: DashboardContainerInput,
    embeddableFactories: EmbeddableFactoryRegistry,
    parent?: Container
  ) {
    super(
      DASHBOARD_CONTAINER_TYPE,
      {
        panels: {},
        isFullScreenMode: false,
        filters: [],
        useMargins: true,
        ...initialInput,
      },
      { embeddableLoaded: {} },
      embeddableFactories,
      parent
    );
  }

  protected createNewPanelState<EEI extends EmbeddableInput = EmbeddableInput>(
    factory: EmbeddableFactory<EEI>,
    partial: Partial<EmbeddableInputMissingFromContainer<EEI, DashboardEmbeddableInput>> & {
      id?: string;
    } = {}
  ): DashboardPanelState<EmbeddableInputMissingFromContainer<EEI, DashboardEmbeddableInput>> {
    const panelState = super.createNewPanelState(factory, partial);
    return createPanelState<EmbeddableInputMissingFromContainer<EEI, DashboardEmbeddableInput>>(
      panelState,
      Object.values(this.input.panels)
    );
  }

  public onToggleExpandPanel(id: string) {
    const newValue = this.input.expandedPanelId ? undefined : id;
    this.updateInput({
      expandedPanelId: newValue,
    });
  }

  public onPanelsUpdated = (panels: { [panelId: string]: DashboardPanelState }) => {
    this.updateInput({
      panels: {
        ...panels,
      },
    });
  };

  public onExitFullScreenMode = () => {
    this.updateInput({
      isFullScreenMode: false,
    });
  };

  public render(dom: React.ReactNode) {
    ReactDOM.render(
      // @ts-ignore
      <I18nProvider>
        <DashboardViewport embeddableFactories={this.embeddableFactories} container={this} />
      </I18nProvider>,
      dom
    );
  }

  public getPanelIndexPatterns() {
    const indexPatterns: IndexPattern[] = [];
    Object.values(this.embeddables).forEach(embeddable => {
      if (!isErrorEmbeddable(embeddable)) {
        const embeddableIndexPatterns = embeddable.getOutput().indexPatterns;
        if (embeddableIndexPatterns) {
          indexPatterns.push(...embeddableIndexPatterns);
        }
      }
    });
    return indexPatterns;
  }

  protected getInheritedInput(id: string): DashboardEmbeddableInput {
    const { viewMode, refreshConfig, timeRange, query, hidePanelTitles, filters } = this.input;
    return {
      filters,
      hidePanelTitles,
      query,
      timeRange,
      refreshConfig,
      viewMode,
      id,
      firstName: 'suuuue',
    };
  }
}
