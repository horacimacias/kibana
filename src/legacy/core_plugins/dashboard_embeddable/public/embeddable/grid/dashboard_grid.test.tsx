/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { shallowWithIntl, nextTick } from 'test_utils/enzyme_helpers';
// @ts-ignore
import sizeMe from 'react-sizeme';

import { ViewMode, EmbeddableFactoryRegistry } from 'plugins/embeddable_api/index';

import { DashboardGrid, DashboardGridProps } from './dashboard_grid';
import {
  HelloWorldEmbeddableFactory,
  HELLO_WORLD_EMBEDDABLE,
} from 'plugins/embeddable_api/__test__';
import { DashboardContainer } from '../dashboard_container';
import { getSampleDashboardInput } from 'plugins/dashboard_embeddable/__test__';

jest.mock('ui/chrome', () => ({ getKibanaVersion: () => '6.0.0' }), { virtual: true });

jest.mock(
  'ui/notify',
  () => ({
    toastNotifications: {
      addDanger: () => {},
    },
  }),
  { virtual: true }
);

let dashboardContainer: DashboardContainer | undefined;

function getProps(props?: Partial<DashboardGridProps>): DashboardGridProps {
  const embeddableFactories = new EmbeddableFactoryRegistry();
  embeddableFactories.registerFactory(new HelloWorldEmbeddableFactory());
  dashboardContainer = new DashboardContainer(
    getSampleDashboardInput({
      panels: {
        '1': {
          gridData: { x: 0, y: 0, w: 6, h: 6, i: '1' },
          embeddableId: '1',
          type: HELLO_WORLD_EMBEDDABLE,
          partialInput: { firstName: 'Bob' },
        },
        '2': {
          gridData: { x: 6, y: 6, w: 6, h: 6, i: '2' },
          type: HELLO_WORLD_EMBEDDABLE,
          embeddableId: '2',
          partialInput: { firstName: 'Stacey' },
        },
      },
    }),
    embeddableFactories
  );
  const defaultTestProps: DashboardGridProps = {
    embeddableFactories,
    container: dashboardContainer,
    intl: null as any,
  };
  return Object.assign(defaultTestProps, props);
}

beforeAll(() => {
  // sizeme detects the width to be 0 in our test environment. noPlaceholder will mean that the grid contents will
  // get rendered even when width is 0, which will improve our tests.
  sizeMe.noPlaceholders = true;
});

afterAll(() => {
  sizeMe.noPlaceholders = false;
});

test('renders DashboardGrid', () => {
  const component = shallowWithIntl(<DashboardGrid.WrappedComponent {...getProps()} />);
  expect(component).toMatchSnapshot();
  const panelElements = component.find('InjectIntl(DashboardPanelUi)');
  expect(panelElements.length).toBe(2);
});

test('renders DashboardGrid with no visualizations', () => {
  const props = getProps();
  props.container.updateInput({ panels: {} });
  const component = shallowWithIntl(<DashboardGrid.WrappedComponent {...props} />);
  expect(component).toMatchSnapshot();
});

test('DashboardGrid removes panel when removed from container', async () => {
  const props = getProps();
  const component = shallowWithIntl(<DashboardGrid.WrappedComponent {...props} />);
  const originalPanels = props.container.getInput().panels;
  const filteredPanels = { ...originalPanels };
  delete filteredPanels['1'];
  props.container.updateInput({ panels: filteredPanels });
  await nextTick();
  component.update();
  const panelElements = component.find('InjectIntl(DashboardPanelUi)');
  expect(panelElements.length).toBe(1);
});
