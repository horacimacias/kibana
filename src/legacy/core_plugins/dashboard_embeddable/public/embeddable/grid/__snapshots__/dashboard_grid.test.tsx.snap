// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardGrid removes panel when removed from container 1`] = `
<SizeMe(ResponsiveGrid)
  isViewMode={true}
  layout={Array []}
  onLayoutChange={[Function]}
/>
`;

exports[`renders DashboardGrid 1`] = `
<SizeMe(ResponsiveGrid)
  isViewMode={true}
  layout={
    Array [
      Object {
        "h": 6,
        "i": "1",
        "w": 6,
        "x": 0,
        "y": 0,
      },
      Object {
        "h": 6,
        "i": "2",
        "w": 6,
        "x": 6,
        "y": 6,
      },
    ]
  }
  onLayoutChange={[Function]}
>
  <div
    className=""
    key="1"
    style={
      Object {
        "zIndex": "auto",
      }
    }
  >
    <InjectIntl(DashboardPanelUi)
      container={
        DashboardContainer {
          "anyChangeListeners": Array [
            [Function],
            [Function],
          ],
          "destoyed": false,
          "embeddableFactories": EmbeddableFactoryRegistry {
            "factories": Object {
              "hello_world": HelloWorldEmbeddableFactory {
                "name": "hello_world",
                "savedObjectMetaData": undefined,
              },
            },
          },
          "embeddables": Object {},
          "id": "123",
          "input": Object {
            "customization": Object {},
            "filters": Array [],
            "id": "123",
            "isFullScreenMode": false,
            "panels": Object {
              "1": Object {
                "embeddableId": "1",
                "gridData": Object {
                  "h": 6,
                  "i": "1",
                  "w": 6,
                  "x": 0,
                  "y": 0,
                },
                "partialInput": Object {
                  "firstName": "Bob",
                },
                "type": "hello_world",
              },
              "2": Object {
                "embeddableId": "2",
                "gridData": Object {
                  "h": 6,
                  "i": "2",
                  "w": 6,
                  "x": 6,
                  "y": 6,
                },
                "partialInput": Object {
                  "firstName": "Stacey",
                },
                "type": "hello_world",
              },
            },
            "query": Object {
              "language": "kuery",
              "query": "hi",
            },
            "timeRange": Object {
              "from": "now-15m",
              "to": "now",
            },
            "title": "My Dashboard",
            "useMargins": false,
            "viewMode": "view",
          },
          "inputChangeListeners": Array [
            [Function],
          ],
          "isContainer": true,
          "onExitFullScreenMode": [Function],
          "onPanelsUpdated": [Function],
          "output": Object {
            "embeddableLoaded": Object {},
          },
          "outputChangeListeners": Array [],
          "panelContainer": undefined,
          "parent": undefined,
          "parentChangesUnsubscribe": undefined,
          "type": "dashboard",
        }
      }
      embeddableId="1"
    />
  </div>
  <div
    className=""
    key="2"
    style={
      Object {
        "zIndex": "auto",
      }
    }
  >
    <InjectIntl(DashboardPanelUi)
      container={
        DashboardContainer {
          "anyChangeListeners": Array [
            [Function],
            [Function],
          ],
          "destoyed": false,
          "embeddableFactories": EmbeddableFactoryRegistry {
            "factories": Object {
              "hello_world": HelloWorldEmbeddableFactory {
                "name": "hello_world",
                "savedObjectMetaData": undefined,
              },
            },
          },
          "embeddables": Object {},
          "id": "123",
          "input": Object {
            "customization": Object {},
            "filters": Array [],
            "id": "123",
            "isFullScreenMode": false,
            "panels": Object {
              "1": Object {
                "embeddableId": "1",
                "gridData": Object {
                  "h": 6,
                  "i": "1",
                  "w": 6,
                  "x": 0,
                  "y": 0,
                },
                "partialInput": Object {
                  "firstName": "Bob",
                },
                "type": "hello_world",
              },
              "2": Object {
                "embeddableId": "2",
                "gridData": Object {
                  "h": 6,
                  "i": "2",
                  "w": 6,
                  "x": 6,
                  "y": 6,
                },
                "partialInput": Object {
                  "firstName": "Stacey",
                },
                "type": "hello_world",
              },
            },
            "query": Object {
              "language": "kuery",
              "query": "hi",
            },
            "timeRange": Object {
              "from": "now-15m",
              "to": "now",
            },
            "title": "My Dashboard",
            "useMargins": false,
            "viewMode": "view",
          },
          "inputChangeListeners": Array [
            [Function],
          ],
          "isContainer": true,
          "onExitFullScreenMode": [Function],
          "onPanelsUpdated": [Function],
          "output": Object {
            "embeddableLoaded": Object {},
          },
          "outputChangeListeners": Array [],
          "panelContainer": undefined,
          "parent": undefined,
          "parentChangesUnsubscribe": undefined,
          "type": "dashboard",
        }
      }
      embeddableId="2"
    />
  </div>
</SizeMe(ResponsiveGrid)>
`;

exports[`renders DashboardGrid with no visualizations 1`] = `
<SizeMe(ResponsiveGrid)
  isViewMode={true}
  layout={Array []}
  onLayoutChange={[Function]}
/>
`;
